### PR TITLE
Enh/precision estimator

### LINKED ISSUE #16: We need to clone a LinearRegressor object here
https://github.com/skmediate/skmediate/blob/4dcceecda7293369a5c74474663494a4a4c0f48d/skmediate/conditional_independence.py#L45

Otherwise, residualization tries to use the same regressor result twice.

### LINKED ISSUE #17: Reverse the order of these
https://github.com/skmediate/skmediate/blob/4dcceecda7293369a5c74474663494a4a4c0f48d/skmediate/conditional_independence.py#L92

The paper has T followed by S, so we should have Y followed by Z.

### PR DESCRIPTION
**Pull request by richford**
## Description

This PR
- reverses the order of Y, Z in the computation of covariance for better accordance with the reference text.
- Clones the input regressors and covariance estimators so that they can be separately fit
- Uses the input precision estimators to compute the `yy` and `zz` precisions. This is especially important for high-dimensional data where one might want to use e.g. `LedoitWolf()` to estimate the precision.
- Adds convenience string parameters for input so that the use doesn't have to import various covariance estimators

## Related Issue

Resolves #16 
Resolves #17

## Type of Change

- [ ] 📚 Examples / docs / tutorials / dependencies update
- [ ] 🔧 Bug fix (non-breaking change which fixes an issue)
- [x] 🥂 Improvement (non-breaking change which improves an existing feature)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🔐 Security fix

## Checklist

- [x] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/mnarayan/skmediate/blob/master/CODE_OF_CONDUCT.md) document.
- [x] I've read the [`CONTRIBUTING.md`](https://github.com/mnarayan/skmediate/blob/master/CONTRIBUTING.md) guide.
- [ ] I've updated the code style using `make codestyle`.
- [ ] I've written tests for all new methods and classes that I created.
- [x] I've written the docstring in Google format for all the methods and classes that I used.

BTW, I can't do `make codestyle` because I can't find the Makefile.

### CODE REVIEW

File: skmediate/conditional_independence.py [hunk 1]
```diff
@@ -81,28 +144,29 @@ def fit(self, X, Z, Y):
...
-        self.prec_zz_ = linalg.pinvh(self.cov_zz_, check_finite=False)
-        self.prec_yy_ = linalg.pinvh(self.cov_yy_, check_finite=False)
+        # Step 2c: Estimate precision of Y
+        self.covfit_yy_ = self.precision_estimator.fit(self.residualized_Y_)
+        self.prec_yy_ = self.covfit_yy_.precision_
 
+        # Step 2d: Calculate residual cross-covariance
         self.residual_crosscovariance_ = (
             (self.cov_yz_ @ self.prec_zz_) @ self.cov_yz_.T
         ) @ self.prec_yy_
```
Comment (arokem) [inline skmediate/conditional_independence.py:172]: Shouldn't this now be:

```suggestion
        self.residual_crosscovariance_ = (
            (self.cov_yz_.T @ self.prec_zz_) @ self.cov_yz_
        ) @ self.prec_yy_
```

to match this eq? 
Screen Shot 2021-07-02 at 10 20 37 AM

**Review by arokem — COMMENTED** [review]
Looks great! I think we need one more change (see suggestion) to match the paper. 

Also, would be great to add tests. Maybe I can do that during the sprint later today.

File: skmediate/conditional_independence.py (hunk 1 above)
Comment (richford) [reply skmediate/conditional_independence.py:172]: I'm confused by this. If  has dimensions , and  has dimensions , then isn't it the case that  has dimensions ,  has dimensions . If so then



is the one that makes sense to me from a dimensional analysis perspective.

**Comment by 36000** [comment]
Whats the status with this PR? Should I try running this on the UKbio data?

**Comment by arokem** [comment]
I'm disinclined to merge this PR, because we should add some testing. 

But I think that it would be great to try this out on the UKBB dataset that you've been analyzing. 

@richford : when you get a chance, would you mind posting a snippet that shows how you did bootstrapping to get confidence intervals on `residual_crosscovariance_corrected_ `?

**Comment by richford** [comment]
Sure! Assuming your `Z` is a white matter bundle in the `wm_bundle` variable and your `Y` is in a variable called `y`, and your `X` is in a variable called `x`. Then the following should do the trick. The below code uses lowess to residualize.

```
from tqdm.auto import tqdm
from sklearn.utils import resample
from skmediate import conditional_independence
from statsmodels.nonparametric.smoothers_lowess import lowess

n_bootstrap = 100

rcc = []
rcc_corrected = []
for n in tqdm(range(n_bootstrap)):
    wm_bs = resample(wm_bundle)
    
    wm_residualized = wm_bs.copy()
    for col_idx, col in enumerate(tqdm(wm_bs.T)):
        wm_residualized[:, col_idx] -= lowess(col, x, return_sorted=False)
    
    ccc = conditional_independence.ConditionalCrossCovariance(
        covariance_estimator=cov_type,
        precision_estimator=cov_type,
        residualized=True,
    )
    ccc.fit(wm_residualized, y[:, np.newaxis])
    rcc.append(ccc.residual_crosscovariance_)
    rcc_corrected.append(ccc.residual_crosscovariance_corrected_)
```

At some point this should go in the docs, I guess.

**Comment by 36000** [comment]
I can do this, but would it make sense to add an optional parameter for doing CV on the regressors or covariance estimators?

**Comment by richford** [comment]
I’d argue that you can always pass in a CV estimator as the regression or covariance estimator. (e.g. pass `LassoCV` for the regressor or `GraphicalLassoCV` for the covariance estimator). And then if you want some custom CV estimator, then yeah one could do that outside the mediation class and pass residualized=True. But that’s just my two cents.

File: skmediate/conditional_independence.py [hunk 2]
```diff
@@ -42,37 +127,76 @@ def __init__(
...
+                    f"{regression_estimator} instead. These instances have the "
+                    f"following method resolution order:\n{mro}"
+                )
             self.regression_estimator_xz = regression_estimator[0]
             self.regression_estimator_xy = regression_estimator[1]
         else:
-            self.regression_estimator_xz = (
-                self.regression_estimator_xy
-            ) = regression_estimator
+            if not isinstance(regression_estimator, RegressorMixin):
```
Comment (36000) [inline skmediate/conditional_independence.py:144]: Maybe we can loosen this up a bit, or add an argument to ignore this error? We really want anything with a fit and predict method. There are multiple useful tools that look like scikit learn models but do not necessarily inherit the RegressorMixin.

**Comment by 36000** [comment]
> Sure! Assuming your `Z` is a white matter bundle in the `wm_bundle` variable and your `Y` is in a variable called `y`, and your `X` is in a variable called `x`. Then the following should do the trick. The below code uses lowess to residualize.
> 
> ```
> from tqdm.auto import tqdm
> from sklearn.utils import resample
> from skmediate import conditional_independence
> from statsmodels.nonparametric.smoothers_lowess import lowess
> 
> n_bootstrap = 100
> 
> rcc = []
> rcc_corrected = []
> for n in tqdm(range(n_bootstrap)):
>     wm_bs = resample(wm_bundle)
>     
>     wm_residualized = wm_bs.copy()
>     for col_idx, col in enumerate(tqdm(wm_bs.T)):
>         wm_residualized[:, col_idx] -= lowess(col, x, return_sorted=False)
>     
>     ccc = conditional_independence.ConditionalCrossCovariance(
>         covariance_estimator=cov_type,
>         precision_estimator=cov_type,
>         residualized=True,
>     )
>     ccc.fit(wm_residualized, y[:, np.newaxis])
>     rcc.append(ccc.residual_crosscovariance_)
>     rcc_corrected.append(ccc.residual_crosscovariance_corrected_)
> ```
> 
> At some point this should go in the docs, I guess.

In this example, does y also have to be residualized?

File: skmediate/conditional_independence.py [hunk 3]
```diff
@@ -81,30 +205,81 @@ def fit(self, X, Z, Y):
...
+            shuffle_cov_est = clone(self.covariance_estimator)
+            shuffle_prec_est = clone(self.precision_estimator)
+
+            if self.show_progress:
+                shuffle_range = trange(self.n_shuffle)
+            else:
+                shuffle_range = range(self.n_shuffle)
+
+            for n in shuffle_range:
+                shuffle_Y = shuffle(Y)
```
Comment (36000) [inline skmediate/conditional_independence.py:263]: I think this should be 
`shuffle_Y = shuffle(self.residualized_Y_)`

Comment (richford) [reply skmediate/conditional_independence.py]: You're right! Thanks! Fixed in latest commit.

File: skmediate/conditional_independence.py (hunk 2 above)
Comment (richford) [reply skmediate/conditional_independence.py:144]: True. We could relax this. One caveat is that we use `sklearn.base.clone` to make sure that the xy and xz regressors don't get mixed up. If we want to keep this, then we have two options:

1. Use `safe=False` in the parameters to `clone`. This will reproduce the current behavior for sklearn estimators and then fall back to a deep copy on objects that are not estimators. Not sure if this have the intended effect.
2. In addition to requiring a `fit` and `predict` method, we also require a `get_params` method so that it works with `clone` without setting `safe=False`. This would work but would place more burden on the user. Personally I would inherit from `RegressorMixin` rather than write my own `get_params` method.

Option 1 is the easiest for us to implement and probably gives the most flexibility to the user.

@arokem, @36000: thoughts?

**Comment by richford** [comment]
@arokem, I got the smoke testing up to 100% coverage. This doesn't involve actual correctness testing, but maybe that's better left for another issue/PR.

File: skmediate/conditional_independence.py (hunk 2 above)
Comment (arokem) [reply skmediate/conditional_independence.py:144]: My thinking is that we would probably gain some clarity from enforcing full compliance with the `RegressorMixin`. Arguably, if a user (even one of us) wants to pass in something that does not inherit from `RegressorMixin`, they should first wrap that thing in an object that does inherit from `RegressorMixin`, make sure that works as expected, and then proceed from there. So, if I had to choose between these two options, I would probably go with 2. Is that too strict?

Comment (richford) [reply skmediate/conditional_independence.py]: I went with 2 in the latest commit.

**Comment by arokem** [comment]
This looks good to me! @36000 : just making sure -- are there any outstanding questions on your end?

**Comment by 36000** [comment]
No outstanding questions, looks ready to merge. Thanks for doing this PR, @richford !

**Comment by arokem** [comment]
Yes: thanks for all your work on this!
